packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layout.tsx

```diff
@@ -161,41 +171,288 @@ const layoutDiagram = (diagram: Diagram) => {
+  return borderNode.data.isBorderNode && borderNode.data.borderNodePosition === BorderNodePositon.WEST;
+};
+const isBorderNodeNorth = (borderNode: Node<NodeData>): boolean => {
+  return borderNode.data.isBorderNode && borderNode.data.borderNodePosition === BorderNodePositon.NORTH;
+};
+const isBorderNodeSouth = (borderNode: Node<NodeData>): boolean => {
+  return borderNode.data.isBorderNode && borderNode.data.borderNodePosition === BorderNodePositon.SOUTH;
+};
+
+const getBorderNodeExtent = (nodeParent: Node<NodeData>, node: Node<NodeData>): CoordinateExtent | 'parent' => {
+  let coordinateExtent: CoordinateExtent | 'parent' = 'parent';
+  if (nodeParent.width && nodeParent.height && node.height && node.width) {
+    switch (node.data.borderNodePosition) {
+      case BorderNodePositon.EAST:
+        coordinateExtent = [
+          [nodeParent.width, 0],
+          [nodeParent.width, nodeParent.height - node.height],
+        ];
+        break;
+      case BorderNodePositon.NORTH:
+        coordinateExtent = [
+          [0 - node.width * 0.8, 0 - node.height],
+          [nodeParent.width - node.width * 0.3, 0 - node.height],
+        ];
+        break;
+      case BorderNodePositon.SOUTH:
+        coordinateExtent = [
+          [0 - node.width * 0.8, nodeParent.height],
+          [nodeParent.width - node.width * 0.3, nodeParent.height],
+        ];
+        break;
+      case BorderNodePositon.WEST:
+        coordinateExtent = [
+          [0 - node.width, 0],
+          [0 - node.width, nodeParent.height - node.height],
+        ];
+        break;
+      default:
+        coordinateExtent = 'parent';
+    }
+  }
+  return coordinateExtent;
+};
+
+const getNodeFootprint = (allVisibleNodes: Node<NodeData>[], nodeChildren: Node<NodeData>): Rect => {
+  const borderNodes = allVisibleNodes
+    .filter((node) => node.parentNode === nodeChildren.id)
+    .filter((node) => node.data.isBorderNode);
+
+  const footPrint: Box = [nodeChildren, ...borderNodes].reduce(
+    (currentFootPrint, node) => {
+      const nodeBox = rectToBox({
+        x: node.data.isBorderNode ? nodeChildren.position.x + node.position.x : node.position.x,
+        y: node.data.isBorderNode ? nodeChildren.position.y + node.position.y : node.position.y,
+        width: node.width ?? 0,
+        height: node.height ?? 0,
+      });
+
+      return getBoundsOfBoxes(currentFootPrint, nodeBox);
+    },
+    { x: Infinity, y: Infinity, x2: -Infinity, y2: -Infinity }
+  );
+  return boxToRect(footPrint);
+};
+
+const getChildNodePosition = (
+  allVisibleNodes: Node<NodeData>[],
+  child: Node<NodeData>,
+  labelElement: HTMLElement | null,
+  previousSibling?: Node<NodeData>
+) => {
+  const maxBorderNodeWidthWest = allVisibleNodes
+    .filter((node) => node.parentNode === child.id)
+    .filter((child) => child.data.isBorderNode && child.data.borderNodePosition === BorderNodePositon.WEST)
+    .map((child) => child.width || 0)
```

Comment:
Should not you base the max border on something like `getFootprint(child).width` instead of `child.width`?